doc/developer/tracing.md

```diff
@@ -0,0 +1,231 @@
+# Introduction
+
+[`tracing`] is a set of crates used for _logging_ and _tracing_ of rust
+binaries. It is primarily composed of the [`tracing`] crate, which is the
+core way to interact with the framework, and the [`tracing-subscriber`] crate,
+which provides an API to compose systems that consume tracing data.
+
+
+## Logging
+
+The most basic way to use `tracing` is to use it like other logging crates like `log`,
+which additional semi-structured capabilities.
+
+
+Materialize binaries are setup such that [events] are printed on stderr.
+
+For example:
+```
+tracing::info!("processed {} items", items.len()")
+```
+produces:
+```
+2022-02-15T18:40:14.289898Z INFO my::module: processed 100 items
+```
+(See [this page for how the colors will look])
+
+You also also add semi-structured data to the event, for example:
+```
+tracing::debug!("processed {} items", items.len(), ?enum_value")
+```
+produces:
+```
+2022-02-15T18:40:14.289898Z DEBUG my::module: processed 100 items _enum_value_=EnumVariant
+```
+because `?field` means record `field` as its `Debug` impl. See [recording fields] for more
+options and examples. This just touches the surface of what you can do with the event macros,
+see [`tracing`] and [`tracing-subscriber`] for more info!
+
+### Levels and Filtering
+
+You may have noticed that log events have [`level`]s. In materialize binaries, we only print
+events that are level `INFO` or more urgent. You can control this level by using the
+`MZ_LOG_FILTER` environment variable to control the level at a _per-crate_, _per-module_ granularity.
+
+For example:
+```
+MZ_LOG_FILTER=my_interesting_module::sub_module=trace,error
+```
+logs ALL events in the specific module, but only `error` events for everything else. See [`Targets`] for more examples.
+
+
+Note that this value is passed through to all compute and storage instances when running locally.
+In the future, it is possible to add per-service filtering, if deemed useful.
+
+## Spans
+
+**Please see [the tracing docs](https://docs.rs/tracing/latest/tracing/index.html#spans) for more information about spans**
+
+[`Span`]s are how `tracing` distinguishes itself from other logging crates. Spans are like events,
```

Comment:
Nitpick, but this makes it sound like spans are something unique to the `tracing` crate, and not a broader o11y concept.